File: LeetCode/1137_NthTribonacciNumber.java
// https://leetcode.cn/problems/n-th-tribonacci-number/
// 2022年10月20日 20:26:38
class Solution {
    public int tribonacci(int n) {
        // v2 dp
        int[] a;
        if (n>2){
            a=new int[n+1];
        }else{
            a=new int[3];
        }
        a[0]=0; a[1]=1;a[2]=1;
        for (int i=3;i<n+1;i++){
            a[i]=a[i-1]+a[i-2]+a[i-3];
        }
        return a[n];
        // // v1 recursion TLE
        // if(n<2){return n;}
        // if(n==2){return 1;}
        // return tribonacci(n-1)+tribonacci(n-2)+tribonacci(n-3);
    }
}

File: LeetCode/1137_NthTribonacciNumber.go
// https://leetcode.cn/problems/n-th-tribonacci-number/
// 2023年03月25日 15:41:20
func tribonacci(n int) int {
	a0 := 0
	a1 := 1
	a2 := 1
	a3 := 0
	if n == 0 {
		return 0
	}
	if n <= 2 {
		return 1
	}
	for i := 2; i < n; i++ {
		a3 = a0 + a1 + a2
		a0 = a1
		a1 = a2
		a2 = a3
	}
	return a3
}

// recurtsive TLE
// 2023年03月25日 15:42:24
func tribonacci(n int) int {
	if n <= 0 {
		return 0
	}
	if n <= 2 {
		return 1
	}
	return tribonacci(n-1) + tribonacci(n-2) + tribonacci(n-3)
}